casper/src/it/scala/coop.rchain.casper/HashSetCasperSpecification.scala

```diff
@@ -0,0 +1,212 @@
+    Sync[Effect].attempt(
+      node.casperEff.addBlock(signed, ignoreDoppelgangerCheck[Effect])
+    )
+
+  def deployment(i: Int, ts: Long = System.currentTimeMillis()): DeployData =
+    ProtoUtil.sourceDeploy(s"new x in { x!(0) }", ts, accounting.MAX_VALUE)
+}
+
+object HashSetCasperSpecification extends Commands {
+  implicit def noShrink[T]: Shrink[T] = Shrink.shrinkAny
+  import HashSetCasperActions._
+
+  type State = List[RNode]
+  type Sut = List[NodeBox]
+
+  def canCreateNewSut(newState: State, initSuts: Traversable[State], runningSuts: Traversable[Sut]): Boolean = {
+    true
+  }
+
+  def initialPreCondition(state: State): Boolean = {
+    true
+  }
+
+  def newSut(state: State): Sut = {
+    val (genesis, validatorKeys) = context(state.size, validators => {
+      val weights = Random.shuffle((1L to validators.size).toList)
+      validators.zip(weights).toMap
+    })
+
+    val network = TestNetwork.empty[Effect]
+
+    val nodes = HashSetCasperTestNode
+      .networkEff(validatorKeys.take(state.size), genesis, testNetwork = network)
+      .map(_.toList)
+
+    nodes.value.unsafeRunSync.right.get.map(new NodeBox(_, None))
+  }
+
+  def destroySut(sut: Sut): Unit = {
+    sut.traverse_(_.node.tearDown()).value.unsafeRunSync.right.get
+  }
+
+  def genInitialState: Gen[State] = for {
+    count: Int <- Gen.chooseNum(2, 4)
+    ids = 0 to count
+    nodes <- Gen.sequence[List[RNode], RNode](ids.map(genRNode))
+  } yield nodes
+
+  def genRNode(i: Int): Gen[RNode] = {
+    Gen.const(RNode(i, s"v$i", deployed = false))
+  }
+
+  def genCommand(state: State): Gen[Command] = for {
+    idx: Int <- Gen.chooseNum(0, state.size - 1)
+    n = state(idx)
+    c <- Gen.frequency(
+      (2, genDeployOrPropose(n)),
+      (1, genReceive(n))
+    )
+  } yield c
+
+  def genDeployOrPropose(n: RNode): Gen[Command] = {
+    if(n.deployed) Gen.const(Propose(n))
+    else Gen.frequency(
+      (1, Gen.const(Deploy(n))),
+      (1, Gen.const(DeployAndPropose(n)))
+    )
+  }
+
+  def genReceive(n: RNode): Gen[Command] = Gen.const(Receive(n))
+
+  case class Deploy(node: RNode) extends Command {
+    type Result = List[String]
+
+    def run(sut: Sut): Result = {
+      val v = sut(node.idx)
+      val rbm = (deploy(v.node, deployment(0)) >> create(v.node)).value.unsafeRunSync.right.get
+      v.update(Some(rbm))
+      v.node.logEff.errors ++ v.node.logEff.warns
+    }
+
+    def nextState(state: State): State = state.patch(node.idx, Seq(node.copy(deployed = true)), 1)
+
+    def preCondition(state: State): Boolean = !state(node.idx).deployed
+
+    def postCondition(state: State, result: Try[Result]): Prop = result.isSuccess && result.get.isEmpty
+
+  }
+
+  case class Propose(node: RNode) extends Command {
+    type Result = (BlockStatus, List[String])
+
+    def run(sut: Sut): Result = {
+      val v = sut(node.idx)
+      val r = add(v.node, v.lastBlock.get).value.unsafeRunSync.right.get.right.get
```

Comment:
lol - can we add an extra val here?

Reply:
And make the `val`s names a bit more verbose